New Post - Secrets modules (#78)
PowerShell/Community-Blog: Posts/2021/10/tfl-secrets.md

---
post_title: How to Use The Secrets Module
username: tfl@psp.co.uk
Categories: PowerShell
tags: PowerShell, profile

> Summary: Using Profile files with PowerShell 7

Author: secrets!"

                                                         machine based on the current user account context. 
                                                         The secrets are encrypted on file using .NETCrypto APIs. 
                                                         A password is required in the default configuration. 
                                                         The configuration can be changed with the provided cmdlets.
                                                         Go to GitHub for more information about this module 
                                                         and to submit issues: https:////github.com//powershell//SecretStore

PS> # 2. Install both modules
PS> Install-Module -Name $Names -Force -AllowClobber
```
When you install the module using `Install-Module` you see no output (unless you use the `-Verbose` switch).
You can always use `Get-Module` to check that you have installed these new (to you) modules.

## Discovering the commands available to you
Once you have thess two modules installed, you can discover the commands in each module:

```powershell-console

PS> # 3. Examine them

> PS> Get-Module -Name Microsoft*.Secret* -List |
>        Format-Table -Property ModuleType*, Ver*, Name, Ps*, Exportedc*

Reviewer: Not sure why you are using so many wildcards here. `PS*` doesn't match anything - the property is **CompatiblePSEditions**. `Exportedc*` matches **ExportedCmdlets** and **ExportedCommands** but only one is displayed here. Also, you might consider using `Format-List` here instead.
```suggestion
PS> Get-Module -Name Microsoft*.Secret* -ListAvailable |
       Format-Table -Property ModuleType, Version, Name, ExportedCmdlets
```

                                                         [Get-SecretInfo, Get-SecretInfo], [Remove-Secret, Remove-Secret],
                                                         [Unlock-SecretVault, Unlock-SecretVault]}
    Binary 1.0.5   Microsoft.PowerShell.SecretStore      {[Unlock-SecretStore, Unlock-SecretStore], [Set-SecretStorePassword,  
                                                         Set-SecretStorePassword], [Get-SecretStoreConfiguration,
                                                         Get-SecretStoreConfiguration], [Set-SecretStoreConfiguration,
                                                         Set-SecretStoreConfiguration], [Reset-SecretStore,
                                                         Reset-SecretStore]}

```

As you can see, both modules have a number of commands you may need to use to manage secrets for your environment.

## Registering and viewing a secret vault

After you have the two modules installed, your next step is to register a secret vault.
There are several vault options you can take advantage of, for this post, I'll use the built-in default vault.
You configure the default vault like this:

```powershell-console

> PS C:\Foo> # 4. Register the default secrets provider

Reviewer: Probably should import the modules first. Not everyone has auto-import enabled.

PS C:\Foo> $Mod = 'Microsoft.PowerShell.SecretStore'
PS C:\Foo> Register-SecretVault -Name RKSecrets -ModuleName $Mod -DefaultVault
PS C:\Foo> Get-SecretVault

Name      ModuleName                       IsDefaultVault
----      ----------                       --------------
RKSecrets Microsoft.PowerShell.SecretStore True
```
Like the previous step, registering the vault does not create any output by default.
You can view the vault you just created by using the `Get-SecretVault` command.

## Setting a secret

> To create a new secret in your secret vault, you use the 1Set-Secret1 command, like this:

Reviewer: ```suggestion
To create a new secret in your secret vault, you use the `Set-Secret` command, like this:
```

```powershell-console

PS> # 4. Register the default secrets provider
PS> $Mod = 'Microsoft.PowerShell.SecretStore'
PS> Register-SecretVault -Name RKSecrets -ModuleName $Mod -DefaultVault
PS> # 5. View Secret vault
PS> Get-SecretVault

Name      ModuleName                       IsDefaultVault
----      ----------                       --------------
RKSecrets Microsoft.PowerShell.SecretStore True

> PS C:\Foo> # 6. Set the Admin password secret for Reskit forest
> PS C:\Foo> Set-Secret -Name ReskitAdmin -Secret 'Pa$$w0rd'

Reviewer: You need to escape the backslash or get rid of the path in in the prompt. Since the path is not relevant, get rid of it.
```suggestion
PS> # 6. Set the Admin password secret for Reskit forest
PS> Set-Secret -Name ReskitAdmin -Secret 'Pa$$w0rd'
```

Creating a new RKSecrets vault. A password is required by the current store configuration.
Enter password:
**********
Enter password again for verification:
**********
```

The first time you use `Set-Secret` to create a secret, the cmdlet prompts for a vault password.
Note this password isd NOT stored in the AD - so don't forget it!!!

## Using secrets stored in your secret vault

Now that you have set a password in the RKSecrets vault, you can use the `Get-Secret` cmdlet to retrieve the secret.
As you can see here, although you set a plain text password, Get-Secret returns the secret as a secure string.

```powershell-console
PS> # 7. Create a credential object using the secet

> PS> $User = 'Reskit\Administrator'

Reviewer: ```suggestion
PS> $User = 'Reskit\\Administrator'
```

PS> $PwSS = Get-Secret ReskitAdmin

> PS> $Cred = [System.Management.Automation.PSCredential]::New($User,$PwSS)

Reviewer: For folks new to these concepts they may not understand what you are doing here. The goal is to convert the username/password pair into a PSCredential object, because that is what the PowerShell cmdlets want to use. And you might want to explain why you need to use the .NET class for this. I know that `Get-Credential` does not allow you to pass in a password, but many people may not.

PS> # 8. Let's cheat and see what the password is first.
PS> $PW = $Cred.GetNetworkCredential().Password
PS> "Password for this credential is [$PW]"
Password for this credential is [Pa$$w0rd]
PS> # 9. Using the credential against DC1
PS> $Cmd = {hostname.exe}
PS> Invoke-Command -ComputerName DC1 -Credential $Cred -ScriptBlock $Cmd
DC1
```

As you can see, it is straightforward to create a new credential object using a password retrieved from the vault.

> You can use the credential object's **GetNetworkCredential()** method to retrieve the plain text password.

Reviewer: ```suggestion
You can use the credential object's `GetNetworkCredential()` method to retrieve the plain text password.
```

The first time you create a vault, the secrets module requires you to specify a vault password.
Depending on what sequence of commands you enter and how quickly, you may be asked to re-enter your vault password.

## Using Metadata

If you have a large numbers of secrets to manage, you can add additional metadata to help you keep track of the secrets you set.
Metadata is a simple hash table containing the metadata you wish to apply to a secret.
Each item in the hash table is a key-value pair.
The keys can be anything you wish such as the purpose of the script and the script author.
You use `Set-Secret` to add metadata to an existing (or new) secret.
To set the metadata, you can use the `Get-SecretInfo` cmdlet.
Creating and using metadata looks like this:

```powershell-console
PS> # 10. Setting metadata

> PS> Set-Secret -Name ReskitAdmin -Secret 'Pa$$w0rd' -Metadata @{Purpose="Reskit.Org Enterprise//Domain Admin PW"}

Reviewer: ```suggestion
PS> Set-Secret -Name ReskitAdmin -Secret 'Pa$$w0rd' -Metadata @{Purpose="Reskit.Org Enterprise\\Domain Admin PW"}
```

PS> Get-SecretInfo -Name ReskitAdmin | Select-Object -Property Name, Metadata

Name        Metadata
----        --------
ReskitAdmin {[Purpose, Reskit.Org Enterprise/Domain Admin PW]}

PS> # 11. Updating the metadata
PS> Set-SecretInfo -Name ReskitAdmin -Metadata @{Author = 'DoctorDNS@Gmail.Com';

>                                              Purpose="Reskit.Org Enterprise//Domain Admin PW"}

Reviewer: ```suggestion
                                             Purpose="Reskit.Org Enterprise\\Domain Admin PW"}
```

PS> # 12. View secret information with metadata
PS> Get-SecretInfo -Name ReskitAdmin | Select-Object -Property Name, Metadata

Name        Metadata
----        --------

> ReskitAdmin {[Purpose, Reskit.Org Enterprise//Domain Admin PW], 

Reviewer: ```suggestion
ReskitAdmin {[Purpose, Reskit.Org Enterprise\\Domain Admin PW], 
```
